django/contrib/admin/decorators.py

```diff
@@ -1,3 +1,71 @@
+def action(function=None, *, permissions=None, description=None):
```

Comment:
> Was just thinking through what this means for ordering of arguments. `function` is a defaulted argument that may be passed positionally _or_ by keyword. So `action(permissions=..., function=...)` worked.

While true, this is intended to be used as a decorator, so the wrapped function will always be passed as the first positional argument. If you try to pass `function=...` to the decorator, it'll fail as the remaining arguments are keyword only. This is also why I didn't mention `function` in the signatures in the documentation, as this is really just an implementation detail to avoid needing to have empty parentheses if no arguments are provided.

> ...making `function` positional only with a default, but that's only available on Python 3.8+.

Actually, I think we'd want `def action(function=None, /, *, permissions=None, description=None):` to enforce that `function` is positional-only and `description` and `permissions` are keyword-only. I think there is benefit to them being keyword only, as it forces them to be explicitly defined. This is particularly important where permissions are concerned. But as you say, waiting for 3.8.

Reply:
> Actually, I think we'd want def action(function=None, /, *, permissions=None, description=None): to enforce that function is positional-only and description and permissions are keyword-only.

Ah yes, my bad! Gosh these separators are a bit tricky :)

Reply:
Here is the explanation I wrote some time ago :)
https://python.astrotech.io/advanced/function/parameter-syntax.html